src/device_memory_tracker.h

```diff
@@ -0,0 +1,265 @@
+    CUptiResult _status = call;                               \
+    if (_status != CUPTI_SUCCESS) {                           \
+      const char* errstr;                                     \
+      cuptiGetResultString(_status, &errstr);                 \
+      LOG_ERROR << #call << " failed with error: " << errstr; \
+      exit(EXIT_FAILURE);                                     \
+    }                                                         \
+  } while (0)
+
+#define BUF_SIZE (32 * 1024)
+#define ALIGN_SIZE (8)
+#define ALIGN_BUFFER(buffer, align)                                 \
+  (((uintptr_t)(buffer) & ((align)-1))                              \
+       ? ((buffer) + (align) - ((uintptr_t)(buffer) & ((align)-1))) \
+       : (buffer))
+
+void
+bufferRequested(uint8_t** buffer, size_t* size, size_t* maxNumRecords)
+{
+  uint8_t* bfr = (uint8_t*)malloc(BUF_SIZE + ALIGN_SIZE);
+  if (bfr != nullptr) {
+    *size = BUF_SIZE;
+    *buffer = ALIGN_BUFFER(bfr, ALIGN_SIZE);
+    *maxNumRecords = 0;
+  } else {
+    LOG_ERROR << "Failed to allocate buffer for CUPIT: out of memory";
+  }
+}
+
+void
+bufferCompleted(
+    CUcontext ctx, uint32_t streamId, uint8_t* buffer, size_t size,
+    size_t validSize)
+{
+  CUptiResult status;
+  CUpti_Activity* record = nullptr;
+
+  if (validSize > 0) {
+    do {
+      status = cuptiActivityGetNextRecord(buffer, validSize, &record);
+      if (status == CUPTI_SUCCESS) {
+        DeviceMemoryTracker::TrackActivity(record);
+      } else if (status == CUPTI_ERROR_MAX_LIMIT_REACHED)
+        break;
+      else {
+        CUPTI_CALL(status);
+      }
+    } while (1);
+
+    // report any records dropped from the queue
+    size_t dropped;
+    CUPTI_CALL(cuptiActivityGetNumDroppedRecords(ctx, streamId, &dropped));
+    if (dropped != 0) {
+      LOG_WARNING << "Dropped " << dropped << " activity records";
+    }
+  }
+
+  free(buffer);
+}
+
+}  // namespace
+
+void
+DeviceMemoryTracker::InitTrace()
+{
+  if (tracker_ == nullptr) {
+    tracker_.reset(new DeviceMemoryTracker());
+    CUPTI_CALL(
+        cuptiActivityRegisterCallbacks(bufferRequested, bufferCompleted));
+    CUPTI_CALL(cuptiActivityEnable(CUPTI_ACTIVITY_KIND_RUNTIME));
+    CUPTI_CALL(cuptiActivityEnable(CUPTI_ACTIVITY_KIND_MEMORY2));
+    CUPTI_CALL(cuptiActivityEnable(CUPTI_ACTIVITY_KIND_EXTERNAL_CORRELATION));
+  }
+}
+
+void
+DeviceMemoryTracker::TrackThreadMemoryUsage(ScopedMemoryUsage* usage)
+{
+  if (tracker_ == nullptr) {
+    InitTrace();
+  }
+  CUPTI_CALL(cuptiActivityPushExternalCorrelationId(
+      CUPTI_EXTERNAL_CORRELATION_KIND_UNKNOWN,
+      reinterpret_cast<uint64_t>(usage)));
+  usage->tracked_ = true;
+}
+
+void
+DeviceMemoryTracker::UntrackThreadMemoryUsage(ScopedMemoryUsage* usage)
+{
+  uint64_t id;
+  if (tracker_ == nullptr) {
+    InitTrace();
+  }
+  CUPTI_CALL(cuptiActivityPopExternalCorrelationId(
+      CUPTI_EXTERNAL_CORRELATION_KIND_UNKNOWN, &id));
+  CUPTI_CALL(cuptiActivityFlushAll(0));
```

Comment:
Can you confirm that this is accurate? Didn't went into details of CUPTI tracing, but I would assume flushing out activities for the popped id above.

Reply:
What do you mean? From what I understand it just flush out the activities recorded, by the time this function is called, all activities has been done and recorded. Let me know if that is not the case.